### PR TITLE
Added Scale subresource support

### LINKED ISSUE #7: Evaluate options for controlling HPA-based scaling
The current implementation of the HVPA controller only controls VPA-based scaling. This is because VPA provides the `updatePolicy: Off` option to switch off VPA updates and only consume recommendations.

Some of the alternatives for a similar controlled HPA-based scaling are below with some preliminary evaluation.

1. Enhance HPA spec with a `updatePolicy` with the functionality similar to VPA.
    - Pros
      - Better/Cleaner API
    - Cons
      - Change in API might take a long time to decide and become available
1. Introduce a `Scale` sub-resource in the HVPA resource to act as a proxy `scaleTargetRef` to the HPA.
    - Pros
      - Can be done already with existing HPA
    - Cons
      - Not so clean API. The dependency/aggregation/composition hierarchy is mixed up.
      - Selectors will have to be replicated from the actual `scaleTargetRef`.

### PR DESCRIPTION
**Pull request by ggaurav10**
**What this PR does / why we need it**:
Added Scale subresource support

**Which issue(s) this PR fixes**:
Fixes #7 

**Special notes for your reviewer**:
I am working on adding test cases. Raising the PR so that the review can happen in parallel

**Release note**:

```noteworthy operator
Added Scale subresource support - Only for HPA `MetricSourceType`s: `Pods` and `Resource`. For `MetricSourceType`s - `Object` and `External` - only `MetricTargetType` of `Value` is supported. This is because these metric sources rely upon `scale.spec.replicas`. Other metric sources rely on `scale.status.replicas`, support for which would be added at later point in time.
```

### CODE REVIEW

File: controllers/hpa_util.go [hunk 1]
```diff
@@ -127,10 +127,14 @@ func getHpaFromHvpa(hvpa *autoscalingv1alpha1.Hvpa) (*autoscaling.HorizontalPodA
...
-			ScaleTargetRef: *hvpa.Spec.TargetRef.DeepCopy(),
-			Metrics:        hvpa.Spec.Hpa.Template.Spec.Metrics,
+			MaxReplicas: hvpa.Spec.Hpa.Template.Spec.MaxReplicas,
+			MinReplicas: hvpa.Spec.Hpa.Template.Spec.MinReplicas,
+			ScaleTargetRef: autoscaling.CrossVersionObjectReference{
+				APIVersion: "autoscaling.k8s.io/v1alpha1",
+				Kind:       "Hvpa",
+				Name:       hvpa.Name,
+			},
+			Metrics: hvpa.Spec.Hpa.Template.Spec.Metrics,
```
Comment (amshuman-kr) [inline controllers/hpa_util.go]: DeepCopy?

File: controllers/hvpa_controller.go [hunk 2]
```diff
@@ -503,14 +502,14 @@ func (r *HvpaReconciler) scaleIfRequired(hpaStatus *autoscaling.HorizontalPodAut
...
+		return nil, false, false, 0, nil, err
 	}
 
 	var desiredReplicas int32
-	if hpaStatus == nil {
+	if hvpa.Spec.Replicas == nil {
 		desiredReplicas = currentReplicas
 	} else {
-		desiredReplicas = hpaStatus.DesiredReplicas
+		desiredReplicas = *hvpa.Spec.Replicas
```
Comment (amshuman-kr) [inline controllers/hvpa_controller.go]: What about hpaWeight?

File: controllers/hvpa_controller.go [hunk 3]
```diff
@@ -693,7 +694,7 @@ func getWeightedReplicas(hpaStatus *autoscaling.HorizontalPodAutoscalerStatus, h
 		return outHpaStatus, err
 	}
 
-	lastScaleTime := hvpa.Status.LastScaling.LastScaleTime.DeepCopy()
+	lastScaleTime := hvpa.Status.LastScaling.LastUpdated.DeepCopy()
 	overrideScaleUpStabilization := hvpa.Status.OverrideScaleUpStabilization
 	if overrideScaleUpStabilization {
```
Comment (amshuman-kr) [inline controllers/hvpa_controller.go]: Shouldn't we still ignore weight in case of `overrideScaleUpStabilization`?

File: controllers/hvpa_controller.go [hunk 4]
```diff
@@ -1281,6 +1283,150 @@ func getThreshold(thresholdVals *autoscalingv1alpha1.ChangeParams, resourceType
...
+		scalingStatus.VpaStatus = podResources
+
+		if podSpec != nil {
+			// If podSpec is passed, we want to capture "Resources" fully for containers for which VPA taget is available.
+			// The podResources.ContainerResources[i] is populated here only for the container for which VPA target is available.
+			for i := range scalingStatus.VpaStatus.ContainerResources {
+				for j := range podSpec.Containers {
+					container := podSpec.Containers[j]
+					if scalingStatus.VpaStatus.ContainerResources[i].ContainerName == container.Name {
+						scalingStatus.VpaStatus.ContainerResources[i].Resources = container.Resources
```
Comment (amshuman-kr) [inline controllers/hvpa_controller.go]: Wouldn't this potentially override [this](https://github.com/gardener/hvpa-controller/pull/65/files#diff-6354ac486a904d10093c497f09f83a3eR1406)?

File: controllers/hvpa_controller.go [hunk 5]
```diff
@@ -1295,9 +1441,15 @@ func getThreshold(thresholdVals *autoscalingv1alpha1.ChangeParams, resourceType
...
 
-	// Fetch the Hvpa instance
+	// Fetch the Hvpa instance using unstructured object so that it is fetched using clientReader, and not the cacheReader
+	hvpaObj := &unstructured.Unstructured{}
+	hvpaObj.SetGroupVersionKind(controllerKindHvpa)
+
+	err := r.Get(ctx, req.NamespacedName, hvpaObj)
 	instance := &autoscalingv1alpha1.Hvpa{}
-	err := r.Get(ctx, req.NamespacedName, instance)
+	if errInt := runtime.DefaultUnstructuredConverter.FromUnstructured(hvpaObj.Object, instance); errInt != nil {
```
Comment (amshuman-kr) [inline controllers/hvpa_controller.go]: Is there no other way to bypass the cache?

**Review by amshuman-kr — CHANGES_REQUESTED** [review]
Overall looks good. Just a few comments and suggestions.

File: controllers/hvpa_controller.go (hunk 2 above)
Comment (ggaurav10) [reply controllers/hvpa_controller.go]: This just captures the HPA's recommendations. The actual calculations based on `hpaWeight` happens later in `getWeightedReplicas()`

File: controllers/hvpa_controller.go (hunk 3 above)
Comment (ggaurav10) [reply controllers/hvpa_controller.go]: The thought had crossed my mind as well, but decided against it :)

Considering following cases:
1. hpaWeight == 0. So vpaWeight == 100. the memory requirement would be fulfilled by a full vertical scaling.
1. hpaWeight > 0. There should be no need to override weight in this case, since both VPA and HPA recommendations will be taken into account
1. If `spec.vpa.deploy` is set to `false`, or the vpa mode is set to `Off`, then it should be upto the user to set memory as a parameter for HPA scaling if he/she encounters any issue

Let me know if this makes sense, or I have missed something.

Comment (amshuman-kr) [reply controllers/hvpa_controller.go]: I was just asking. Makes sense.

File: controllers/hvpa_controller.go (hunk 4 above)
Comment (ggaurav10) [reply controllers/hvpa_controller.go]: Yes, this is the idea here.
Here we want to capture full details only for those containers for which VPA has provided recommendations (in its `status.target`). Please note that VPA recommendations can be turned off individually for containers.
Since vpa's `target` doesn't have `limits`, we depend on `podSpec`, and overwrite what was previously captured using `target`.
This enables hvpa-controller to control `resources` only for containers for which VPA is enabled, so that other containers can still be controlled from outside, ie, can be manually changed and will not be reconciled using hvpa's `lastApplied`

File: controllers/hvpa_controller.go (hunk 5 above)
Comment (ggaurav10) [reply controllers/hvpa_controller.go]: Doesn't look like it. :/

File: controllers/hvpa_controller.go (hunk 4 above)
Comment (amshuman-kr) [reply controllers/hvpa_controller.go]: But then it might container the resources for only the current VPA recommendation which may not recommend for every container. Isn't [`ProcessedRecommendations`](https://github.com/gardener/hvpa-controller/pull/65/files#diff-61c5bea64511bdaec1e64fb43fc1773cR282) supposed to contain the latest resources for all the containers so far recommended by VPA. Not just what came from the current recommendation?

Comment (ggaurav10) [reply controllers/hvpa_controller.go]: All the recommendations are kept in VPA's status. Even if there are new recommendations for specific containers in the status, the recommendations of other containers are not removed from the VPA's status. So we always have all - changed and unchanged - recommendations all the time in VPA's status.

> But then it might contain the resources for only the current VPA recommendation which may not recommend for every container.

Even if this is a possibility that for certain containers VPA recommendations were never populated in the VPA's status until a time, then there is anyway no way for hvpa-controller to capture them, or to control the `resources` of those containers. The recommendations for those containers will be captured as and when they are updated in VPA's status. This will be like turning the update mode of those specific container from "Off" to "Auto".
Also, as mentioned above, VPA keeps the older recommendations of other containers in its status.
I hope this clarifies.

File: controllers/hpa_util.go (hunk 1 above)
Comment (ggaurav10) [reply controllers/hpa_util.go]: the hvpa instance passed here is anyway a local instance, and doesn't point to the cache. So didn't bother about deep-copying. wdyt?

File: controllers/hvpa_controller.go (hunk 4 above)
Comment (amshuman-kr) [reply controllers/hvpa_controller.go:1428]: 👍 I was thinking we could get the potentially missing values from the target pod template. But what you mentioned is fine.

File: controllers/hpa_util.go (hunk 1 above)
Comment (amshuman-kr) [reply controllers/hpa_util.go:136]: Might be safer to DeepCopy because it is now being pointed to from HPA and HVPA. Just to be safe.

File: controllers/hvpa_controller.go [hunk 6]
```diff
@@ -527,7 +527,7 @@ func (r *HvpaReconciler) scaleIfRequired(hpaStatus *autoscaling.HorizontalPodAut
 		log.Error(err, "Error in getting weight based requests in new deployment", "hvpa", hvpa.Namespace+"/"+hvpa.Name)
 	}
 
-	hpaStatus, err = getWeightedReplicas(hpaStatus, hvpa, currentReplicas, 100-vpaWeight, blockedScaling)
+	hpaStatus, err = getWeightedReplicas(hpaStatus, hvpa, currentReplicas, hpaWeight, blockedScaling)
```
Comment (amshuman-kr) [inline controllers/hvpa_controller.go:530]: `hpaWeight` here would have been `100-vpaWeight` but now might be `MaxWeight`. Is this ok?

File: controllers/hvpa_controller.go [hunk 7]
```diff
@@ -979,7 +986,9 @@ func getWeightedRequests(vpaStatus *vpa_api.VerticalPodAutoscalerStatus, hvpa *a
 					blockedByWeight = true
 
 				} else if diffMem.Sign() > 0 {
-					if hpaScaleOutLimited == false || scaleUpUpdateMode == autoscalingv1alpha1.UpdateModeOff {
+					// Optimization: If we are fully dependent on VPA, then let's wait for HPA scale out to be limited before scaling up
+					// TODO: Should we have a dedicated BlockingReason for this case?
+					if (hpaScaleOutLimited == false && vpaWeight == autoscalingv1alpha1.MaxWeight) || scaleUpUpdateMode == autoscalingv1alpha1.UpdateModeOff {
```
Comment (amshuman-kr) [inline controllers/hvpa_controller.go:991]: Do we need this case now anymore? When would this happen?

File: controllers/hvpa_controller.go [hunk 8]
```diff
@@ -1041,7 +1050,9 @@ func getWeightedRequests(vpaStatus *vpa_api.VerticalPodAutoscalerStatus, hvpa *a
 					blockedByWeight = true
 
 				} else if diffCPU.Sign() > 0 {
-					if hpaScaleOutLimited == false || scaleUpUpdateMode == autoscalingv1alpha1.UpdateModeOff {
+					// Optimization: If we are fully dependent on VPA, then let's wait for HPA scale out to be limited before scaling up
+					// TODO: Should we have a dedicated BlockingReason for this case?
+					if (hpaScaleOutLimited == false && vpaWeight == autoscalingv1alpha1.MaxWeight) || scaleUpUpdateMode == autoscalingv1alpha1.UpdateModeOff {
```
Comment (amshuman-kr) [inline controllers/hvpa_controller.go:1055]: Do we need this case now anymore? When would this happen?

File: controllers/hvpa_controller_test.go [hunk 9]
```diff
@@ -217,11 +218,34 @@ var _ = Describe("#TestReconcile", func() {
...
 			Eventually(func() bool {
-				h := &autoscalingv1alpha1.Hvpa{}
+				h := &hvpav1alpha1.Hvpa{}
 				c.Get(context.TODO(), types.NamespacedName{Name: instance.Name, Namespace: instance.Namespace}, h)
 				return h.Status.OverrideScaleUpStabilization
 			}, timeout).Should(BeTrue())
 
+			// Update VPA status, let HVPA scale
+			hvpa := &hvpav1alpha1.Hvpa{}
+			Expect(c.Get(context.TODO(), types.NamespacedName{Name: instance.Name, Namespace: instance.Namespace}, hvpa)).To(Succeed())
```
Comment (amshuman-kr) [inline controllers/hvpa_controller_test.go:228]: Can we add some additional test cases to cover the new scenarios because of the new control over HPA scaling?

**Review by amshuman-kr — CHANGES_REQUESTED** [review]
Mostly look good. Some questions. Also can we add additional test cases to cover the edge cases of the support for control over HPA.

File: controllers/hvpa_controller.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ggaurav10) [reply controllers/hvpa_controller.go:530]: Fixed in last update

File: controllers/hvpa_controller.go (hunk 7 above)
Comment (ggaurav10) [reply controllers/hvpa_controller.go:991]: As mentioned in the comment, this is an optimisation so that vertical scaling happens only when HPA is limited. We can still continue having this, atleast until we tweak hvpa spec for our apiservers. WDYT?

File: controllers/hvpa_controller.go [hunk 10]
```diff
@@ -785,14 +787,18 @@ func getWeightedReplicas(hpaStatus *autoscaling.HorizontalPodAutoscalerStatus, h
 	return outHpaStatus, err
 }
 
-func isHpaScaleOutLimited(hpaStatus *autoscaling.HorizontalPodAutoscalerStatus, maxReplicas int32, hpaScaleUpUpdateMode *string, maintenanceWindow *autoscalingv1alpha1.MaintenanceTimeWindow) bool {
+func isHpaScaleOutLimited(hpaStatus *autoscaling.HorizontalPodAutoscalerStatus, hpaWeight autoscalingv1alpha1.VpaWeight, maxReplicas int32, hpaScaleUpUpdateMode *string, maintenanceWindow *autoscalingv1alpha1.MaintenanceTimeWindow) bool {
```
[the remark this answers is not in the document]
Comment (ggaurav10) [reply controllers/hvpa_controller.go:790]: we need this until we have [this](https://github.com/gardener/hvpa-controller/pull/65/files#diff-6354ac486a904d10093c497f09f83a3eR989) optimisation.

**Comment by ggaurav10** [comment]
> can we add additional test cases to cover the edge cases of the support for control over HPA

Added in https://github.com/gardener/hvpa-controller/pull/65/commits/b7ce106d5d7a3a07e30ee6c5d3b31b683973bc6b

File: controllers/hvpa_controller.go (hunk 7 above)
Comment (amshuman-kr) [reply controllers/hvpa_controller.go:991]: Not sure, I understood fully.

Do you mean until we tweak the HVPA spec template for the shoot apiservers in gardener to change the `weightBasedScalingIntervals` to scale vertically first? 

Or do you mean that we should change the HVPASpec type to avoid this case?

Comment (ggaurav10) [reply controllers/hvpa_controller.go:991]: This:
> until we tweak the HVPA spec template for the shoot apiservers in gardener to change the weightBasedScalingIntervals to scale vertically first

Until we do this or any other modification which is helpful, IMO, we can continue using this optimisation the same way we have been. I think it can be useful, and we can even think of keeping it as an unofficial contract (atleast in the beginning) using alpha annotations. Hint: https://github.com/gardener/gardener/pull/2073

Comment (amshuman-kr) [reply controllers/hvpa_controller.go:991]: > Until we do this or any other modification which is helpful, IMO, we can continue using this optimisation the same way we have been. I think it can be useful, and we can even think of keeping it as an unofficial contract (atleast in the beginning) using alpha annotations. Hint: gardener/gardener#2073

Things were getting clear until the hint part which confused me further :-)

**Comment by amshuman-kr** [comment]
> Added in [b7ce106](https://github.com/gardener/hvpa-controller/commit/b7ce106d5d7a3a07e30ee6c5d3b31b683973bc6b)

Thanks for the tests you have added already. What I meant was that we don't have good coverage for the different possible usage of `weightBasedScalingIntervals`. What we [do have](https://github.com/gardener/hvpa-controller/blob/b7ce106d5d7a3a07e30ee6c5d3b31b683973bc6b/controllers/suite_test.go#L239) in the tests is something that we will probably not use for shoot apiservers in gardener (now or with `Scale` sub-resource support).

Can you please add some variations for this which cover the current usage of `weightBasedScalingIntervals` in gardener as well as what we might use when we integrate `Scale` sub-resource support with gardener?

File: controllers/hvpa_controller.go (hunk 7 above)
Comment (ggaurav10) [reply controllers/hvpa_controller.go:991]: I meant that instead of completely removing this optimisation, we can enable it iff some annotation is set on the hvpa resource. Since its a deviation from HVPA controller's definition/contract, we can enable it only if a user has added an annotation. WDYT

Comment (amshuman-kr) [reply controllers/hvpa_controller.go:991]: > Hint: [gardener/gardener#2073](https://github.com/gardener/gardener/pull/2073)

How is the annotation support relevant here?

Comment (ggaurav10) [reply controllers/hvpa_controller.go:991]: i mentioned this just as an example how Gardener uses these kind of annotations.. Nothing to do with this particular usage.. :)

Comment (amshuman-kr) [reply controllers/hvpa_controller.go:991]: > I meant that instead of completely removing this optimisation, we can enable it iff some annotation is set on the hvpa resource. Since its a deviation from HVPA controller's definition/contract, we can enable it only if a user has added an annotation. WDYT

I am fine keeping it (possibly even without an annotation). But my question was really the following.

How can `hpaScaleOutLimited == false && vpaWeight == autoscalingv1alpha1.MaxWeight` ever be `true` after this PR? Wouldn't that be an error? Am I missing something?

Comment (ggaurav10) [reply controllers/hvpa_controller.go:991]: Example:
When replicas has reached `maxReplicas` but is not `limited` because HPA wouldn't want to scale out anyway (deducted from `hpa.status.conditions`), but VPA wants to scale up vertically.
Under this `if` condition we block this scale up but only if VPA weight is `MaxWeight`.
It's basically the same behaviour that we have kept for apiserver until now.

**Comment by amshuman-kr** [comment]
@ggaurav10 can we close this in favour of #76 ?

**Comment by ggaurav10** [comment]
yes.. we can /close this.
